### PR TITLE
Adaptive lambdas

### PR DESCRIPTION
**Pull request by nikiniki1**
Realization of adaptive lambdas and dynamic loss function for boundary conditions.

### CODE REVIEW

File: tedeous/metrics.py [hunk 1]
```diff
@@ -260,76 +269,70 @@ def b_op_val_calc(self, bcond) -> torch.Tensor:
...
-                   lambda_bound * torch.mean((b_val - true_b_val) ** 2)
+            loss = torch.mean((self.operator) ** 2) + self.loss_bcs()
         else:
-            loss = torch.sum(torch.mean((op) ** 2, 0)) + \
-                   lambda_bound * torch.sum(torch.mean((b_val - true_b_val) ** 2, 0))
+            loss = torch.sum(torch.mean((self.operator) ** 2, 0)) + self.loss_bcs()
         return loss
 
-    def casual_loss(self, lambda_bound:  Union[int, float] = 10, tol: float = 0) -> torch.Tensor:
+    def casual_loss(self, tol: float = 0) -> torch.Tensor:
```
Comment (SuperSashka) [inline tedeous/metrics.py:335]: А зачем его вычислять можно пояснить?

File: tedeous/metrics.py [hunk 2]
```diff
@@ -371,37 +368,86 @@ def weak_loss(self, weak_form: Union[None, list], lambda_bound: Union[int, float
...
-        b_val, true_b_val = self.bconds.apply_bconds_operator()
-
-        loss = torch.sum(op) + \
-               lambda_bound * torch.sum(torch.mean((b_val - true_b_val) ** 2, 0))
+        loss = torch.sum(self.operator) + self.loss_bcs()
 
         return loss
 
-    def loss_evaluation(self, lambda_bound: Union[int, float] = 10, weak_form: Union[None, list] = None, tol=0) -> Union[default_loss, weak_loss, casual_loss]:
+    def compute(self, tol=0) -> \
```
Comment (SuperSashka) [inline tedeous/metrics.py:380]: tol влияет на штраф, можно тоже всюду пояснить, как влияет-то?

File: tedeous/metrics.py [hunk 3]
```diff
@@ -371,37 +368,86 @@ def weak_loss(self, weak_form: Union[None, list], lambda_bound: Union[int, float
...
+        self.model = model.to(device_type())
+        self.mode = mode
+        self.weak_form = weak_form
+        self.lambda_bound = lambda_bound
+        self.operator = Operator(self.grid, self.prepared_operator, self.model,
+                                   self.mode, weak_form)
+        self.boundary = Bounds(self.grid, self.prepared_bconds, self.model,
+                                   self.mode, weak_form)
+
+    def evaluate(self, iter: int, update_every_lambdas: Union[None, int], tol: float)-> torch.Tensor:
```
Comment (SuperSashka) [inline tedeous/metrics.py:429]: Если это выносится пользователю, то мб лучше kwargs использовать? .evaluate(-1, None, 0) так себе информативно.

**Review by SuperSashka — CHANGES_REQUESTED** [review]
Надо поработать над понятностью кода сначала. Все примеры ещё не запускал.

**Comment by SuperSashka** [comment]
What we have done so far -  we fully revamped adaprive lambdas routine such that it is computed using dispersion part directly with Sobol indices, not neural tangent kernel eigenvalues analogue. It is done since NTK does not work for anything exept single PDE - in NTK case we would have left ODE and systems out.

Secondly, we rework the loss - it is now computed in two faces - one with lambdas for gradient descent and one with normalized for stopping crtiterion. Even though it is a bit of pulls back everything - namely, training process is not quite connected with stop criterion - it would be benefit for parameter unification.

Additionally, we split Dirichlet and initial (we made a step further and split Dirichlet, operator, periodic) boundary conditions in terms of lambda like big guys. Adaptive lambdas are also split. 

I hope we making last second fixes and pull new shiny feautes asap.

**Review by SuperSashka — APPROVED** [review]
It seems that everything is worked as planned at the timebeing, old functionality is preserved and the new one - adaptive lambdas - works as intended.

It should be noted that the stop criterion work is changed, so it may be required to change work with parameters eps, patience and abs_loss in existing code. The tutorial on solver tuning is on its way.
